### PR TITLE
Add a --no-pin option to `ipfs add`

### PR DESCRIPTION
**Pull request by eminence**
Implements a solution for #1908

This PR replaces #1909

>

### CODE REVIEW

**Comment by lgierth** [comment]
I still think this should maybe be called `--pin` instead, defaulting to true.

**Comment by eminence** [comment]
Totally reasonable.  I'll try to defend the code as implemented by pointing out that seems to be precedent in ipfs for options that, when present, turn off some functionality: 

* We have `ipfs add --only-hash` instead of `ipfs add --write=false`

(Ok, I admit that's the only example I could find).  I don't think I care either way, except that `--no-pin` seems to be easier to type than `--pin=false`.  I am happy to change this if someone else votes in favor of @lgierth

**Comment by jbenet** [comment]
i think in _this_ case, we do want `--pin=false`.

**Comment by eminence** [comment]
Ok!  But one small problem -- In the `dev0.4.0` branch, there is no support for having bool options take true/false parameters.  For example:

```
 $ ./ipfs add --pin=false
Error: Option 'pin' takes no arguments, but was passed 'true'
```

This was added in 107409ce, but this commit is not part of the `dev0.4.0` branch.  Are there any plans to rebase this branch on master?

**Comment by rht** [comment]
+1 `dev0.4.0` rebase @whyrusleeping (this can be done with less risk than other PRs)

**Comment by jbenet** [comment]
@eminence @whyrusleeping is rebasing dev0.4.0 shortly

**Comment by eminence** [comment]
Ok, I force pushed a commit, due to the dev0.4.0 rebase.  Please take another look.  Thanks!

**Comment by whyrusleeping** [comment]
This LGTM

File: core/commands/add.go [hunk 1]
```diff
@@ -94,6 +96,11 @@ remains to be implemented.
 		hash, _, _ := req.Option(onlyHashOptionName).Bool()
 		hidden, _, _ := req.Option(hiddenOptionName).Bool()
 		chunker, _, _ := req.Option(chunkerOptionName).String()
+		dopin, pin_found, _ := req.Option(pinOptionName).Bool()
+
+		if !pin_found { // default
+			dopin = true
+		}
```
Comment (jbenet) [inline core/commands/add.go]: is there a way to set the default on the commands lib itself? (am not sure.)

**Comment by jbenet** [comment]
This LGTM. i'll merge. we can always improve https://github.com/ipfs/go-ipfs/pull/1931#discussion_r44955132 later

File: core/commands/add.go (hunk 1 above)
Comment (rht) [inline core/commands/add.go]: There is the js trick `!!dopin`, don't know if it works in golang.

**Pull request merged**